compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -236,137 +142,640 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                        // This works because they force a cycle, so each row is guaranteed to also be a column
+                        let mut idxs: Vec<usize> =
+                            args.iter().filter(|a| a.is_some()).map(|a| a.unwrap()).collect();
+                        // FIXME: Is there a cleaner way to do this?
+                        let mut real_idxs = vec![None; provided_args.len()];
+                        for (src, dst) in args.iter().enumerate() {
+                            real_idxs[arg_indexes[src]] = dst.map(|dst| arg_indexes[dst]);
+                        }
+                        issues.push(Issue::Permutation(real_idxs));
+                        idxs.sort();
+                        idxs.reverse();
+                        for i in idxs {
+                            satisfy_input(
+                                &mut compatibility_matrix,
+                                &mut input_indexes,
+                                &mut arg_indexes,
+                                i,
+                                i,
+                            );
+                        }
+                    }
+                    None => {
+                        // We didn't find any issues, so we need to push the algorithm forward
+                        // First, eliminate any arguments that currently satisfy their inputs
+                        let mut i = cmp::min(arg_indexes.len(), input_indexes.len());
+                        while i > 0 {
+                            let idx = i - 1;
+                            if compatibility_matrix[idx][idx] {
+                                satisfy_input(
+                                    &mut compatibility_matrix,
+                                    &mut input_indexes,
+                                    &mut arg_indexes,
+                                    idx,
+                                    idx,
+                                );
+                            }
+                            i -= 1;
+                        }
+                    }
+                };
+            }
+
+            let issue_count = issues.len();
+            if issue_count > 0 {
+                // We found issues, so lets construct a diagnostic that summarizes the issues we found
+                // FIXME: This might need some refining in code review
+                let mut labels = vec![];
+                let mut suggestions = vec![];
+                let source_map = self.sess().source_map();
+                for issue in issues {
+                    match issue {
+                        Issue::Invalid(arg) => {
+                            let span = provided_args[arg].span;
+                            let expected_type = expected_input_tys[arg];
+                            let found_type = final_arg_types.iter().filter(|(i, _, _)| *i == arg).next().unwrap().1;
```

Comment:
```suggestion
                            let found_type = final_arg_types.iter().find(|(i, _, _)| *i == arg).unwrap().1;
```

Reply:
For these, I will probably end up changing how final_arg_types is represented, so we're not doing a bunch of scans to find the appropriate argument, but good tip!